### PR TITLE
[SPARK-40209][SQL] Don't change the interval value of Decimal in `changePrecision()` on errors

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to modify the internal value of Decimal `longVal` in the method `changePrecision` only in the success cases, and don't touch it on any errors to keep the original value.

### Why are the changes needed?
To don't confuse users by error messages. The fix might improve user experience with Spark SQL.

### Does this PR introduce _any_ user-facing change?
Yes. The PR changes user-facing errors.

Before:
```sql
spark-sql> select cast(interval '10.123' second as decimal(1, 0));
[NUMERIC_VALUE_OUT_OF_RANGE] 0.000010 cannot be represented ...
```

After:
```sql
spark-sql> select cast(interval '10.123' second as decimal(1, 0));
[NUMERIC_VALUE_OUT_OF_RANGE] 10.123000 cannot be represented ...
```

### How was this patch tested?
By running the affected test suites:
```
$ build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z cast.sql"
```

### CODE REVIEW

**Review by srielau — APPROVED** [review]
LGTM

**Comment by cloud-fan** [comment]
> This change seems to break the design.

I don't think so. Having side effects for better performance is OK, but it doesn't mean we can leave the decimal value in a "crashed" state. `changePrecision` should be a no-op if it returns false.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 1]
```diff
@@ -394,48 +394,49 @@ final class Decimal extends Ordered[Decimal] with Serializable {
...
-        if (diff <= MAX_LONG_DIGITS && longVal > -p && longVal < p) {
-          // Multiplying longVal by POW_10(diff) will still keep it below MAX_LONG_DIGITS
-          longVal *= POW_10(diff)
+        if (diff <= MAX_LONG_DIGITS && lv > -p && lv < p) {
+          // Multiplying lv by POW_10(diff) will still keep it below MAX_LONG_DIGITS
+          lv *= POW_10(diff)
         } else {
           // Give up on using Longs; switch to BigDecimal, which we'll modify below
-          decimalVal = BigDecimal(longVal, _scale)
+          decimalVal = BigDecimal(lv, _scale)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:439]: shall we avoid updating `decimalVal` as well if we eventually return false?

**Comment by MaxGekk** [comment]
> This change seems to break the design.

The bad design is to leave a Decimal value in inconsistent/wrong state between invokes of Decimal methods. Seems like the current assumption is `changePrecision()` can be invoked ONLY once (since we cannot predict either the method fails or not).

**Review by cloud-fan — APPROVED** [review]

**Review by gengliangwang — APPROVED** [review]

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @srielau @gengliangwang @cloud-fan for review.

**Pull request closed without merging**